src/Stratis.Bitcoin/P2P/PeerAddressManager.cs

```diff
@@ -0,0 +1,313 @@
+        {
+            foreach (var networkAddress in networkAddresses)
+            {
+                AddPeer(networkAddress, source);
+            }
+        }
+
+        public void AddPeer(PeerAddress peer)
+        {
+            //[NBitcoin] what is this check?
+            if (peer.NetworkAddress.Endpoint.Address.IsRoutable(true) == false)
+                return;
+
+            this.peers.TryAdd(peer.NetworkAddress.Endpoint.GetHashCode(), peer);
+        }
+
+        public void DeletePeer(PeerAddress peer)
+        {
+            IPEndPoint endPoint = null;
+            this.peers.TryRemove(endPoint.GetHashCode(), out peer);
+        }
+
+        public void PeerAttempted(IPEndPoint endpoint, DateTimeOffset peerAttemptedAt)
+        {
+            var peer = FindPeer(endpoint.Address.ToString(), endpoint.Port);
+            if (peer == null)
+                return;
+
+            peer.Attempted(peerAttemptedAt);
+        }
+
+        public void PeerConnected(IPEndPoint endpoint, DateTimeOffset peerConnectedAt)
+        {
+            var peer = FindPeer(endpoint.Address.ToString(), endpoint.Port);
+            if (peer == null)
+                return;
+
+            peer.Connected(peerConnectedAt);
+        }
+
+        public void PeerHandshaked(IPEndPoint endpoint, DateTimeOffset peerHandshakedAt)
+        {
+            var peer = FindPeer(endpoint.Address.ToString(), endpoint.Port);
+            if (peer == null)
+                return;
+
+            peer.Handshaked(peerHandshakedAt);
+        }
+
+        public PeerAddress FindPeer(string ip, int port)
+        {
+            var peer = this.peers.SingleOrDefault(p => p.Value.Match(ip, port));
+            if (peer.Value != null)
+                return peer.Value;
+            return null;
+        }
+
+        /// <summary>
+        /// Selects a random peer to connect to.
+        /// 
+        /// [NBitcoin] Use a 50% chance for choosing between tried and new table entries.
+        /// </summary>
+        public NetworkAddress SelectPeerToConnectTo()
+        {
+            if (this.peers.Tried().Any() == true &&
+                (this.peers.New().Any() == false || GetRandomInteger(2) == 0))
+                return this.peers.Tried().Random().NetworkAddress;
+
+            if (this.peers.New().Any() == true)
+                return this.peers.New().Random().NetworkAddress;
+
+            return null;
+        }
+
+        /// <summary>
+        /// Selects a random set of preferred peers to connects to.
+        /// </summary>
+        public IEnumerable<NetworkAddress> SelectPeersToConnectTo()
+        {
+            return this.peers.Where(p => p.Value.Preferred).Select(p => p.Value.NetworkAddress);
+        }
+
+        internal static int GetRandomInteger(int max)
+        {
+            return (int)(RandomUtils.GetUInt32() % (uint)max);
+        }
+
+        internal void DiscoverPeers(Network network, NodeConnectionParameters nodeParameters, int peersToFind)
```

Comment:
Lets pull this to a separate class maybe `AddressDiscoveryLoop`

Reply:
I've moved all of the discovery logic to a new class called PeerAddressDiscoveryLoop